openedx/core/djangoapps/monitoring_utils/utils.py

```diff
@@ -0,0 +1,479 @@
+IGNORE_THRESHOLDS = {
+    'ApdexStats': 10,
+    'SplitResult': 50,
+    'TimeStats': 500,
+    'set': 10000,
+}
+
+WAFFLE_NAMESPACE = 'monitoring_utils'
+
+log = logging.getLogger(__name__)
+
+
+def show_memory_leaks(
+        label=u'memory_leaks',
+        max_console_rows=MAX_CONSOLE_ROWS,
+        max_graphed_object_types=MAX_GRAPHED_OBJECT_TYPES,
+        refs_depth=REFS_DEPTH,
+        back_refs_depth=BACK_REFS_DEPTH,
+        max_objects_per_type=MAX_OBJECTS_PER_TYPE,
+        ignore_thresholds=None,
+        graph_directory_path=GRAPH_DIRECTORY_PATH,
+        memory_table_buffer=None,
+        skip_first_graphs=True):
+    """
+    Call this function to get data about memory leaks; what objects are being
+    leaked, where did they come from, and what do they contain?  The leaks
+    are measured from the last call to ``get_new_ids()`` (which is called
+    within this function).  Some data is printed to stdout, and more details
+    are available in graphs stored at the paths printed to stdout.  Subsequent
+    calls with the same label are indicated by an increasing index in the
+    filename.
+
+    Args:
+        label (unicode): The start of the filename for each graph
+        max_console_rows (int): The max number of object types for which to
+            show data on the console
+        max_graphed_object_types (int): The max number of object types for
+            which to generate reference graphs
+        refs_depth (int): Maximum depth of forward reference graphs
+        back_refs_depth (int): Maximum depth of backward reference graphs
+        max_objects_per_type (int): Max number of objects per type to use as
+            starting points in the reference graphs
+        ignore_thresholds (dict): Object type names for which table rows and
+            graphs should not be generated if the new object count is below
+            the corresponding number.
+        graph_directory_path (unicode): The directory in which graph files
+            will be created.  It will be created if it doesn't already exist.
+        memory_table_buffer (StringIO): Storage for the generated table of
+            memory statistics.  Ideally, create this before starting to
+            count newly allocated objects.
+        skip_first_graphs (bool): True if the first call to this function for
+            a given label should not produce graphs (the default behavior).
+            The first call to a given block of code often initializes an
+            assortment of objects which aren't really leaked memory.
+    """
+    if ignore_thresholds is None:
+        ignore_thresholds = IGNORE_THRESHOLDS
+    if memory_table_buffer is None:
+        memory_table_buffer = StringIO()
+    new_ids = get_new_ids(limit=max_console_rows, ignore_thresholds=ignore_thresholds,
+                          output=memory_table_buffer)
+    memory_table_text = memory_table_buffer.getvalue()
+    log.info('\n' + memory_table_text)
```

Comment:
Silly nit/question: Why the pre-pended `\n`?

Reply:
It's because of our evil intentions to ruin log formatting!  The table text is already multi-line, so we start with a newline to prevent losing the first line in the sea of logs.